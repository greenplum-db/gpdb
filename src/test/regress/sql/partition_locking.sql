-- Test locking behaviour. When creating, dropping, querying or adding indexes
-- partitioned tables, we want to lock only the master, not the children.

-- start_ignore
create view locktest as
select coalesce(
  case when relname like 'pg_toast%index' then 'toast index'
  	   when relname like 'pg_toast%' then 'toast table'
	   else relname end, 'dropped table'), 
mode,
locktype from 
pg_locks l left outer join pg_class c on (l.relation = c.oid),
pg_database d where relation is not null and l.database = d.oid and 
l.gp_segment_id = -1 and
(c.relname is null or 
 c.relname like 'locktest_t1%' or
 c.relname like 'pg_toast%index' or
 c.relname like 'pg_toast%') and
-- XXX XXX: ignore gp_fault_strategy and pg_class*, plan is changed, more slices are parallel executed 
-- on entry db, so pg_class* relations are not stable.
d.datname = current_database() order by 1, 3, 2;
-- end_ignore

-- Partitioned table with toast table
begin;

-- creation
create table locktest_t1 (i int, t text) partition by range(i)
(start(1) end(10) every(1));

-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;

-- drop
begin;
drop table locktest_t1;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;

-- AO table (ao segments, block directory won't exist after create)
begin;
-- creation
create table locktest_t1 (i int, t text, n numeric)
with (appendonly = true)
partition by list(i)
(values(1), values(2), values(3));
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;
begin;

-- add a little data
insert into locktest_t1 values(1), (2), (3);
insert into locktest_t1 values(1), (2), (3);
insert into locktest_t1 values(1), (2), (3);
insert into locktest_t1 values(1), (2), (3);
insert into locktest_t1 values(1), (2), (3);
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;

commit;
-- drop
begin;
drop table locktest_t1;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;

-- Indexing
create table locktest_t1 (i int, t text) partition by range(i)
(start(1) end(10) every(1));

begin;
create index locktest_t1_idx on locktest_t1(i);
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;

-- test select locking
begin;
select * from locktest_t1 where i = 1;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;

begin;
-- insert locking
insert into locktest_t1 values(3, 'f');
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;

-- delete locking
begin;
delete from locktest_t1 where i = 4;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;

-- drop index
begin;
drop table locktest_t1;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
commit;
