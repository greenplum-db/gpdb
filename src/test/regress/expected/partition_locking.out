-- Test locking behaviour. When creating, dropping, querying or adding indexes
-- partitioned tables, we want to lock only the master, not the children.
-- start_ignore
create view locktest as
select coalesce(
  case when relname like 'pg_toast%index' then 'toast index'
  	   when relname like 'pg_toast%' then 'toast table'
	   else relname end, 'dropped table'), 
mode,
locktype from 
pg_locks l left outer join pg_class c on (l.relation = c.oid),
pg_database d where relation is not null and l.database = d.oid and 
l.gp_segment_id = -1 and
(c.relname is null or 
 c.relname like 'locktest_t1%' or
 c.relname like 'pg_toast%index' or
 c.relname like 'pg_toast%') and
-- XXX XXX: ignore gp_fault_strategy and pg_class*, plan is changed, more slices are parallel executed 
-- on entry db, so pg_class* relations are not stable.
d.datname = current_database() order by 1, 3, 2;
-- end_ignore
-- Partitioned table with toast table
begin;
-- creation
create table locktest_t1 (i int, t text) partition by range(i)
(start(1) end(10) every(1));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_1" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_2" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_3" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_4" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_5" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_6" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_7" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_8" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_9" for table "locktest_t1"
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
  coalesce   |        mode         | locktype 
-------------+---------------------+----------
 toast index | AccessExclusiveLock | relation
 locktest_t1 | AccessExclusiveLock | relation
 toast table | ShareLock           | relation
(3 rows)

commit;
-- drop
begin;
drop table locktest_t1;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
   coalesce    |        mode         | locktype 
---------------+---------------------+----------
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
(3 rows)

commit;
-- AO table (ao segments, block directory won't exist after create)
begin;
-- creation
create table locktest_t1 (i int, t text, n numeric)
with (appendonly = true)
partition by list(i)
(values(1), values(2), values(3));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_1" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_2" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_3" for table "locktest_t1"
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
  coalesce   |        mode         | locktype 
-------------+---------------------+----------
 locktest_t1 | AccessExclusiveLock | relation
 toast table | ShareLock           | relation
 toast index | AccessExclusiveLock | relation
(3 rows)

commit;
begin;
-- add a little data
insert into locktest_t1 values(1), (2), (3);
insert into locktest_t1 values(1), (2), (3);
insert into locktest_t1 values(1), (2), (3);
insert into locktest_t1 values(1), (2), (3);
insert into locktest_t1 values(1), (2), (3);
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
      coalesce       |        mode         |         locktype         
---------------------+---------------------+--------------------------
 locktest_t1         | RowExclusiveLock    | relation
 locktest_t1_1_prt_1 | AccessExclusiveLock | append-only segment file
 locktest_t1_1_prt_1 | AccessShareLock     | relation
 locktest_t1_1_prt_2 | AccessExclusiveLock | append-only segment file
 locktest_t1_1_prt_2 | AccessShareLock     | relation
 locktest_t1_1_prt_3 | AccessExclusiveLock | append-only segment file
 locktest_t1_1_prt_3 | AccessShareLock     | relation
(7 rows)

commit;
-- drop
begin;
drop table locktest_t1;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
   coalesce    |        mode         | locktype 
---------------+---------------------+----------
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
(7 rows)

commit;
-- Indexing
create table locktest_t1 (i int, t text) partition by range(i)
(start(1) end(10) every(1));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_1" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_2" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_3" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_4" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_5" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_6" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_7" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_8" for table "locktest_t1"
NOTICE:  CREATE TABLE will create partition "locktest_t1_1_prt_9" for table "locktest_t1"
begin;
create index locktest_t1_idx on locktest_t1(i);
NOTICE:  building index for child partition "locktest_t1_1_prt_1"
NOTICE:  building index for child partition "locktest_t1_1_prt_2"
NOTICE:  building index for child partition "locktest_t1_1_prt_3"
NOTICE:  building index for child partition "locktest_t1_1_prt_4"
NOTICE:  building index for child partition "locktest_t1_1_prt_5"
NOTICE:  building index for child partition "locktest_t1_1_prt_6"
NOTICE:  building index for child partition "locktest_t1_1_prt_7"
NOTICE:  building index for child partition "locktest_t1_1_prt_8"
NOTICE:  building index for child partition "locktest_t1_1_prt_9"
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
    coalesce     |        mode         | locktype 
-----------------+---------------------+----------
 locktest_t1     | ShareLock           | relation
 locktest_t1_idx | AccessExclusiveLock | relation
(2 rows)

commit;
-- test select locking
begin;
select * from locktest_t1 where i = 1;
 i | t 
---+---
(0 rows)

-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
      coalesce       |      mode       | locktype 
---------------------+-----------------+----------
 locktest_t1         | AccessShareLock | relation
 locktest_t1_1_prt_1 | AccessShareLock | relation
(2 rows)

commit;
begin;
-- insert locking
insert into locktest_t1 values(3, 'f');
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
  coalesce   |       mode       | locktype 
-------------+------------------+----------
 locktest_t1 | RowExclusiveLock | relation
(1 row)

commit;
-- delete locking
begin;
delete from locktest_t1 where i = 4;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
      coalesce       |     mode      | locktype 
---------------------+---------------+----------
 locktest_t1_1_prt_4 | ExclusiveLock | relation
 locktest_t1         | ExclusiveLock | relation
(2 rows)

commit;
-- drop index
begin;
drop table locktest_t1;
-- start_ignore
-- Known_opt_diff: MPP-20936
-- end_ignore
select * from locktest;
   coalesce    |        mode         | locktype 
---------------+---------------------+----------
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
 dropped table | AccessExclusiveLock | relation
(4 rows)

commit;
